src/librustc/ty/layout.rs

```diff
@@ -555,7 +555,7 @@ impl<'a, 'gcx, 'tcx> Struct {
             }
 
             // Is this the NonZero lang item wrapping a pointer or integer type?
-            (&Univariant { non_zero: true, .. }, &ty::TyStruct(def, substs)) => {
+            (&Univariant { non_zero: true, .. }, &ty::TyAdt(def, substs)) if def.is_struct() => {
```

Comment:
The guard is unnecessary (because `non_zero` means it's the `NonZero` lang item).